### PR TITLE
Code modification for  testcases of various network models in directory example

### PR DESCRIPTION
**Pull request by luobao-intel**
## Description ##
In the example directory, there are some test cases that exits some small bugs and missed options  running  on CPU.  Thie pull request is to add the CPU options and fix some small bugs.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

#### add CPU option ####
- [x] bayesian-methods
- [x] fcn-xs
- [x] multi-task
- [x] rcnn
- [x] rnn-time-major
#### small bugs fix ####
- [x] rcnn
- [x] cnn_visualization
#### file location change ####
- [x] module

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

@pengzhao-intel @TaoLv @juliusshufan

### CODE REVIEW

**Comment by kalyc** [comment]
@luobao-intel thanks for your contribution. Could you update the PR title to be inline with what exactly this is fixing?

@mxnet-label-bot[pr-awaiting-testing]

**Comment by vrakesh** [comment]
@luobao-intel Requesting an update on the PR, any progress on the build failures?

**Comment by luobao-intel** [comment]
@sandeep-krishnamurthy The code has been modified according to your requirements. Can you review it again? Thank you!

File: example/bayesian-methods/bdk_demo.py [hunk 1]
```diff
@@ -180,18 +183,18 @@ def run_mnist_SGLD(training_num=50000):
...
+    exe, sample_pool = SGLD(sym=net, dev=dev(xpu), data_inputs=data_inputs, X=X, Y=Y,
                             X_test=X_test, Y_test=Y_test,
                             total_iter_num=1000000,
                             initializer=initializer,
                             learning_rate=4E-6, prior_precision=1.0, minibatch_size=100,
                             thin_interval=100, burn_in_iter_num=1000)
 
 
-def run_mnist_DistilledSGLD(training_num=50000):
+def run_mnist_DistilledSGLD(training_num=50000, xpu=0):
```
Comment (sandeep-krishnamurthy) [inline example/bayesian-methods/bdk_demo.py]: This means, by default, mx.gpu(0). Should it default to CPU instead by making xpu=None?

**Comment by ankkhedia** [comment]
@luobao-intel 
Thanks for updating the PR. There are few minor review comments. Could you please take a look?

**Comment by kalyc** [comment]
ping! @luobao-intel could you please take a look?

**Comment by juliusshufan** [comment]
@pengxin99

**Comment by juliusshufan** [comment]
@sandeep-krishnamurthy Could you take a look if your comments have been addressed with latest changes? Thanks.

File: example/bayesian-methods/bdk_demo.py (hunk 1 above)
Comment (pengxin99) [reply example/bayesian-methods/bdk_demo.py]: @sandeep-krishnamurthy  The code has been modified according to your requirements. Could you review it again? Thank you!

File: example/bayesian-methods/bdk_demo.py [hunk 2]
```diff
@@ -156,19 +156,22 @@ def get_toy_sym(teacher=True, teacher_noise_precision=None):
     return net
 
 
-def dev():
-    return mx.gpu()
+def dev(xpu):
+    if xpu == None:
+        return mx.cpu()
+    else:
+        return mx.gpu(xpu)
```
Comment (vandanavk) [inline example/bayesian-methods/bdk_demo.py]: maybe 
```
def dev(xpu=None):
    return mx.gpu(xpu) if xpu else mx.cpu()
```
?

File: example/bayesian-methods/bdk_demo.py [hunk 3]
```diff
@@ -156,19 +156,22 @@ def get_toy_sym(teacher=True, teacher_noise_precision=None):
...
 
 
 def run_mnist_SGD(training_num=50000):
     X, Y, X_test, Y_test = load_mnist(training_num)
     minibatch_size = 100
     net = get_mnist_sym()
     data_shape = (minibatch_size,) + X.shape[1::]
-    data_inputs = {'data': nd.zeros(data_shape, ctx=dev()),
-                   'softmax_label': nd.zeros((minibatch_size,), ctx=dev())}
+    data_inputs = {'data': nd.zeros(data_shape, ctx=dev(xpu)),
```
Comment (vandanavk) [inline example/bayesian-methods/bdk_demo.py]: does it need `def run_mnist_SGD(training_num=50000, xpu=None)` here?

File: example/bayesian-methods/bdk_demo.py [hunk 4]
```diff
@@ -180,18 +183,18 @@ def run_mnist_SGLD(training_num=50000):
...
```
Comment (vandanavk) [inline example/bayesian-methods/bdk_demo.py]: `def run_mnist_SGLD(training_num=50000, xpu=None)` here?

File: example/fcn-xs/fcn_xs.py [hunk 5]
```diff
@@ -28,9 +28,9 @@
 
 logger = logging.getLogger()
 logger.setLevel(logging.INFO)
-ctx = mx.gpu(0)
 
 def main():
+    ctx = mx.cpu() if args.gpu == None else mx.gpu(0)
```
Comment (vandanavk) [inline example/fcn-xs/fcn_xs.py]: `if not args.gpu` ?

File: example/rcnn/test.py [hunk 6]
```diff
@@ -35,7 +35,7 @@ def test_net(sym, imdb, args):
     logger.info('called with args\n{}'.format(pprint.pformat(vars(args))))
 
     # setup context
-    ctx = mx.gpu(args.gpu)
+    ctx = mx.cpu() if args.gpu == None else mx.gpu(args.gpu)
```
Comment (vandanavk) [inline example/rcnn/test.py]: if not args.gpu

File: example/rcnn/train.py [hunk 7]
```diff
@@ -33,7 +33,7 @@ def train_net(sym, roidb, args):
     logger.info('called with args\n{}'.format(pprint.pformat(vars(args))))
 
     # setup multi-gpu
-    ctx = [mx.gpu(int(i)) for i in args.gpus.split(',')]
+    ctx = [mx.cpu()] if args.gpus == None else [mx.gpu(int(i)) for i in args.gpus.split(',')]
```
Comment (vandanavk) [inline example/rcnn/train.py]: if not args.gpu

**Comment by pengxin99** [comment]
@vandanavk @sandeep-krishnamurthy  I have change the code reference to the review comments, could you please make a review again? Thanks

**Comment by vandanavk** [comment]
LGTM

**Comment by roywei** [comment]
@sandeep-krishnamurthy could you take a look again? Thanks!

**Comment by pengxin99** [comment]
@vandanavk @roywei @sandeep-krishnamurthy Thanks for your reviews, and could this be merged?

**Comment by pengzhao-intel** [comment]
@sandeep-krishnamurthy could you help take a look again? Thanks.

File: example/bayesian-methods/bdk_demo.py [hunk 8]
```diff
@@ -156,42 +156,42 @@ def get_toy_sym(teacher=True, teacher_noise_precision=None):
     return net
 
 
-def dev():
-    return mx.gpu()
+def dev(xpu = None):
```
Comment (TaoLv) [inline example/bayesian-methods/bdk_demo.py]: No need to have white spaces before and after `=`. Also, `xpu` is a little confusing for me as its value should be either `gpu` or `cpu` according to the name. But here in this case it equals to gpu numbers.

File: example/rcnn/symdata/loader.py [hunk 9]
```diff
@@ -129,7 +129,7 @@ def getpad(self):
 
 class AnchorLoader(mx.io.DataIter):
     def __init__(self, roidb, batch_size, short, max_size, mean, std,
-                 feat_sym, anchor_generator: AnchorGenerator, anchor_sampler: AnchorSampler,
+                 feat_sym, anchor_generator=AnchorGenerator, anchor_sampler=AnchorSampler,
```
Comment (TaoLv) [inline example/rcnn/symdata/loader.py]: Can you explain?

File: example/rcnn/test.py [hunk 10]
```diff
@@ -94,7 +94,7 @@ def parse_args():
     parser.add_argument('--params', type=str, default='', help='path to trained model')
     parser.add_argument('--dataset', type=str, default='voc', help='training dataset')
     parser.add_argument('--imageset', type=str, default='', help='imageset splits')
-    parser.add_argument('--gpu', type=int, default=0, help='gpu device eg. 0')
+    parser.add_argument('--gpu', type=int, default=0, help='gpu device eg. 0 if None then use cpu')
```
Comment (TaoLv) [inline example/rcnn/test.py]: if not provide,  then use cpu.

File: example/rcnn/train.py [hunk 11]
```diff
@@ -127,7 +127,7 @@ def parse_args():
     parser.add_argument('--pretrained', type=str, default='', help='path to pretrained model')
     parser.add_argument('--dataset', type=str, default='voc', help='training dataset')
     parser.add_argument('--imageset', type=str, default='', help='imageset splits')
-    parser.add_argument('--gpus', type=str, default='0', help='gpu devices eg. 0,1')
+    parser.add_argument('--gpus', type=str, help='gpu devices eg. 0,1 if null then use cpu')
```
Comment (TaoLv) [inline example/rcnn/train.py]: if not provide, then use cpu.

File: example/fcn-xs/fcn_xs.py [hunk 12]
```diff
@@ -85,6 +85,7 @@ def main():
         help='the init type of fcn-xs model, e.g. vgg16, fcnxs')
     parser.add_argument('--retrain', action='store_true', default=False,
         help='true means continue training.')
+    parser.add_argument("--gpu", type=int, help="if None then use cpu else use gpu")
```
Comment (TaoLv) [inline example/fcn-xs/fcn_xs.py]: ditto

File: example/fcn-xs/fcn_xs.py [hunk 13]
```diff
@@ -28,9 +28,9 @@
 
 logger = logging.getLogger()
 logger.setLevel(logging.INFO)
-ctx = mx.gpu(0)
 
 def main():
+    ctx = mx.cpu() if not args.gpu else mx.gpu(0)
```
Comment (TaoLv) [inline example/fcn-xs/fcn_xs.py]: mx.gpu(args.gpu)? What's the value of args.gpu?

File: example/rcnn/symdata/loader.py (hunk 9 above)
Comment (pengxin99) [reply example/rcnn/symdata/loader.py]: The type of parameter anchor_generator must by Class AnchorGenerator, so i think it should be `:`, not `=` , i will do changes corresponding

File: example/bayesian-methods/bdk_demo.py [hunk 14]
```diff
@@ -253,20 +253,20 @@ def run_toy_SGLD():
...
     student_net = get_toy_sym(False)
     data_shape = (minibatch_size,) + X.shape[1::]
-    teacher_data_inputs = {'data': nd.zeros(data_shape, ctx=dev()),
-                           'teacher_output_label': nd.zeros((minibatch_size, 1), ctx=dev())}
-    student_data_inputs = {'data': nd.zeros(data_shape, ctx=dev())}
-    #                   'softmax_label': nd.zeros((minibatch_size, 10), ctx=dev())}
+    teacher_data_inputs = {'data': nd.zeros(data_shape, ctx=dev(gpu_id)),
+                           'teacher_output_label': nd.zeros((minibatch_size, 1), ctx=dev(gpu_id))}
+    student_data_inputs = {'data': nd.zeros(data_shape, ctx=dev(gpu_id))}
+    #                   'softmax_label': nd.zeros((minibatch_size, 10), ctx=dev(xpu))}
```
Comment (TaoLv) [inline example/bayesian-methods/bdk_demo.py]: Can be removed.

File: example/bayesian-methods/bdk_demo.py [hunk 15]
```diff
@@ -350,21 +350,22 @@ def run_synthetic_SGLD():
                         help="Type of algorithm to use. 0 --> SGD, 1 --> SGLD, other-->DistilledSGLD")
     parser.add_argument("-t", "--training", type=int, default=50000,
                         help="Number of training samples")
+    parser.add_argument("--gpu", type=int, help="if None then use cpu else use gpu")
```
Comment (TaoLv) [inline example/bayesian-methods/bdk_demo.py]: if None -> if not provided.

**Comment by TaoLv** [comment]
@sandeep-krishnamurthy Please check if your comments are addressed.

**Comment by TaoLv** [comment]
Thank you for the contribution @pengxin99. I think it's good to merge now.
